File: libraries/mcerebrumapi/android/src/main/java/org/md2k/mcerebrumapi/datakitapi/datasource/MCDataSource.java
package org.md2k.mcerebrumapi.datakitapi.datasource;
/*
 * Copyright (c) 2016, The University of Memphis, MD2K Center

 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;

import com.alibaba.fastjson.JSON;

import org.md2k.mcerebrumapi.data.MCDataType;
import org.md2k.mcerebrumapi.datakitapi.datasource.metadata.MCApplicationMetaData;
import org.md2k.mcerebrumapi.datakitapi.datasource.metadata.MCDataDescriptor;
import org.md2k.mcerebrumapi.datakitapi.datasource.metadata.MCDataSourceMetaData;
import org.md2k.mcerebrumapi.datakitapi.datasource.metadata.MCPlatformAppMetaData;
import org.md2k.mcerebrumapi.datakitapi.datasource.metadata.MCPlatformMetaData;

import java.util.ArrayList;
import java.util.HashMap;

public class MCDataSource implements Parcelable {
    private static final String SEPARATOR = "-";

    String dataSourceType = null;
    String dataSourceId = null;
    String platformType = null;
    String platformId = null;
    String platformAppType = null;
    String platformAppId = null;
    String applicationType = null;
    String applicationId = null;


    int dataType = MCDataType.INT_ARRAY.getValue();

    HashMap<String, String> dataSourceMetaData = new HashMap<>();
    HashMap<String, String> platformMetaData = new HashMap<>();
    private HashMap<String, String> platformAppMetaData = new HashMap<>();
    HashMap<String, String> applicationMetaData = new HashMap<>();
    ArrayList<HashMap<String, String>> dataDescriptors = new ArrayList<>();

    protected MCDataSource() {
    }

    protected MCDataSource(String uuid) {
        String[] splits = uuid.split(SEPARATOR);
        if (splits.length > 0 && splits[0] != null && splits[0].length() != 0)
            this.dataSourceType = splits[0];
        if (splits.length > 1 && splits[1] != null && splits[1].length() != 0)
            this.dataSourceId = splits[1];
        if (splits.length > 2 && splits[2] != null && splits[2].length() != 0)
            this.platformType = splits[2];
        if (splits.length > 3 && splits[3] != null && splits[3].length() != 0)
            this.platformId = splits[3];
        if (splits.length > 4 && splits[4] != null && splits[4].length() != 0)
            this.platformAppType = splits[4];
        if (splits.length > 5 && splits[5] != null && splits[5].length() != 0)
            this.platformAppId = splits[5];
        if (splits.length > 6 && splits[6] != null && splits[6].length() != 0)
            this.applicationType = splits[6];
        if (splits.length > 7 && splits[7] != null && splits[7].length() != 0)
            this.applicationId = splits[7];

    }

    public static final Creator<MCDataSource> CREATOR = new Creator<MCDataSource>() {
        @Override
        public MCDataSource createFromParcel(Parcel in) {
            return new MCDataSource(in);
        }

        @Override
        public MCDataSource[] newArray(int size) {
            return new MCDataSource[size];
        }
    };

    public static IDataSourceBuilder.IAppInfo registerBuilder() {
        return new DataSourceRegisterBuilder();
    }

    public static IDataSourceBuilder.IQuery queryBuilder() {
        return new DataSourceQueryBuilder();
    }

    public String getDataSourceType() {
        return dataSourceType;
    }

    public String getDataSourceId() {
        return dataSourceId;
    }

    public String getPlatformType() {
        return platformType;
    }

    public String getPlatformId() {
        return platformId;
    }

    public String getPlatformAppType() {
        return platformAppType;
    }

    public String getPlatformAppId() {
        return platformAppId;
    }

    public String getApplicationType() {
        return applicationType;
    }

    public String getApplicationId() {
        return applicationId;
    }

    public MCDataType getDataType() {
        return MCDataType.getDataType(dataType);
    }

    public MCDataSourceMetaData getDataSourceMetaData() {
        return MCDataSourceMetaData.builder().setMetaData(dataSourceMetaData).build();
    }

    public MCPlatformMetaData getPlatformMetaData() {
        return MCPlatformMetaData.builder().setMetaData(platformMetaData).build();
    }

    public MCPlatformAppMetaData getPlatformAppMetaData() {
        return MCPlatformAppMetaData.builder().setMetaData(platformAppMetaData).build();
    }

    public MCApplicationMetaData getApplicationMetaData() {
        return MCApplicationMetaData.builder().setMetaData(applicationMetaData).build();
    }

    public ArrayList<MCDataDescriptor> getDataDescriptors() {
        ArrayList<MCDataDescriptor> dds = new ArrayList<>();
        for (int i = 0; i < dataDescriptors.size(); i++) {
            dds.add(MCDataDescriptor.builder().setDescriptor(dataDescriptors.get(i)).build());
        }
        return dds;
    }

    @NonNull
    @Override
    public String toString() {
        return JSON.toJSONString(this);
    }

    private MCDataSource(Parcel in) {
        dataSourceType = in.readString();
        dataSourceId = in.readString();
        platformType = in.readString();
        platformId = in.readString();
        platformAppType = in.readString();
        platformAppId = in.readString();
        applicationType = in.readString();
        applicationId = in.readString();

        dataType = in.readInt();
        dataType = in.readInt();

        dataSourceMetaData = readHashMapFromParcel(in);
        platformMetaData = readHashMapFromParcel(in);
        platformAppMetaData = readHashMapFromParcel(in);
        applicationMetaData = readHashMapFromParcel(in);
        dataDescriptors = readHashMapArrayFromParcel(in);

    }


    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeString(dataSourceType);
        parcel.writeString(dataSourceId);
        parcel.writeString(platformType);
        parcel.writeString(platformId);
        parcel.writeString(platformAppType);
        parcel.writeString(platformAppId);
        parcel.writeString(applicationType);
        parcel.writeString(applicationId);

        parcel.writeInt(dataType);
        parcel.writeInt(dataType);

        writeHashMapToParcel(parcel, dataSourceMetaData);
        writeHashMapToParcel(parcel, platformMetaData);
        writeHashMapToParcel(parcel, platformAppMetaData);
        writeHashMapToParcel(parcel, applicationMetaData);
        writeDataDescriptorToParcel(parcel, dataDescriptors);
    }

    private ArrayList<HashMap<String, String>> readHashMapArrayFromParcel(Parcel in) {
        ArrayList<HashMap<String, String>> dataDescriptors;
        int size = in.readInt();
        dataDescriptors = new ArrayList<>();
        for (int i = 0; i < size; i++)
            dataDescriptors.add(readHashMapFromParcel(in));
        return dataDescriptors;
    }

    private void writeDataDescriptorToParcel(Parcel parcel, ArrayList<HashMap<String, String>> dataDescriptors) {
        int size = dataDescriptors.size();
        parcel.writeInt(size);
        for (HashMap<String, String> dataDescriptor : dataDescriptors)
            writeHashMapToParcel(parcel, dataDescriptor);
    }

    private HashMap<String, String> readHashMapFromParcel(Parcel in) {
        HashMap<String, String> metaData = new HashMap<>();
        int size = in.readInt();
        for (int j = 0; j < size; j++) {
            metaData.put(in.readString(), in.readString());
        }
        return metaData;
    }

    private void writeHashMapToParcel(Parcel parcel, HashMap<String, String> metaData) {
        int size = metaData.size();
        parcel.writeInt(size);
        for (HashMap.Entry<String, String> entry : metaData.entrySet()) {
            parcel.writeString(entry.getKey());
            parcel.writeString(entry.getValue());
        }
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof MCDataSource))
            return false;
        MCDataSource d = (MCDataSource) obj;
        return toUUID().equals(d.toUUID())
                && dataSourceMetaData.equals(d.dataSourceMetaData)
                && platformMetaData.equals(d.platformMetaData)
                && platformAppMetaData.equals(d.platformAppMetaData)
                && applicationMetaData.equals(d.applicationMetaData)
                && dataDescriptors.equals(d.dataDescriptors)
                && dataType == d.dataType;
    }

    public String toUUID() {
        return prepString(dataSourceType) + SEPARATOR
                + prepString(dataSourceId) + SEPARATOR
                + prepString(platformType) + SEPARATOR
                + prepString(platformId) + SEPARATOR
                + prepString(platformAppType) + SEPARATOR
                + prepString(platformAppId) + SEPARATOR
                + prepString(applicationType) + SEPARATOR
                + prepString(applicationId);

    }

    private String prepString(String s) {
        if (s == null) return "";
        else return s;
    }

    public boolean isSubset(MCDataSource master) {
        return (this.dataSourceType == null || this.dataSourceType.equals(master.dataSourceType))
                && (this.dataSourceId == null || this.dataSourceId.equals(master.dataSourceId))
                && (this.platformType == null || this.platformType.equals(master.platformType))
                && (this.platformId == null || this.platformId.equals(master.platformId))
                && (this.platformAppType == null || this.platformAppType.equals(master.platformAppType))
                && (this.platformAppId == null || this.platformAppId.equals(master.platformAppId))
                && (this.applicationType == null || this.applicationType.equals(master.applicationType))
                && (this.applicationId == null || this.applicationId.equals(master.applicationId));
    }

    public boolean isEqualUUID(MCDataSource d) {
        return toUUID().equals(d.toUUID());
    }

}

File: libraries/mcerebrumapi/lib/datakitapi/mc_datasource.dart
import 'package:equatable/equatable.dart';
import 'package:flutter/material.dart';
import 'package:package_info/package_info.dart';

import 'mc_data_format.dart';
import 'mc_metadata.dart';

class MCDataSource extends Equatable {
  static const String _SEPARATOR = "-";

  final String dataSourceType;
  final String dataSourceId;
  final String platformType;
  final String platformId;
  final String platformAppType;
  final String platformAppId;
  final String applicationType;
  final String applicationId;

  final MCDataSourceMetaData dataSourceMetaData;
  final MCPlatformMetaData platformMetaData;
  final MCPlatformAppMetaData platformAppMetaData;
  final MCApplicationMetaData applicationMetaData;
  final List<MCDataDescriptor> dataDescriptors;

  final MCDataFormat dataFormat;

  MCDataSource._internal(
      {this.dataSourceType,
      this.dataSourceId,
      this.platformType,
      this.platformId,
      this.platformAppType,
      this.platformAppId,
      this.applicationType,
      this.applicationId,
        this.dataFormat,
      this.dataSourceMetaData,
      this.platformMetaData,
      this.platformAppMetaData,
      this.applicationMetaData,
      this.dataDescriptors});

  static MCDataSource query(
      {String dataSourceType,
      String dataSourceId,
      String platformType,
      String platformId,
      String platformAppType,
      String platformAppId,
      String applicationType,
      String applicationId}) {
    return new MCDataSource._internal(
        dataSourceType: dataSourceType,
        dataSourceId: dataSourceId,
        platformType: platformType,
        platformId: platformId,
        platformAppType: platformAppType,
        platformAppId: platformAppId,
        applicationType: applicationType,
        applicationId: applicationId);
  }

  static Future<MCDataSource> create({
    @required String dataSourceType,
    String dataSourceId,
    String platformType,
    String platformId,
    String platformAppType,
    String platformAppId,
    String applicationId,
    MCDataFormat dataFormat=MCDataFormat.DATA,
    @required MCDataSourceMetaData dataSourceMetaData,
    MCPlatformMetaData platformMetaData,
    MCPlatformAppMetaData platformAppMetaData,
    @required List<MCDataDescriptor> dataDescriptors,
  }) async {
    MCApplicationMetaData applicationMetaData =
        await MCApplicationMetaData.create();
    PackageInfo packageInfo = await PackageInfo.fromPlatform();
    return new MCDataSource._internal(
        dataSourceType: dataSourceType,
        dataSourceId: dataSourceId,
        platformType: platformType,
        platformId: platformId,
        platformAppType: platformAppType,
        platformAppId: platformAppId,
        applicationType: packageInfo.packageName,
        applicationId: applicationId,
        dataSourceMetaData: dataSourceMetaData,
        platformMetaData: platformMetaData,
        platformAppMetaData: platformAppMetaData,
        applicationMetaData: applicationMetaData,
        dataDescriptors: dataDescriptors);
  }

  bool isSubsetOf(MCDataSource dataSource) {
    if (dataSourceType != null &&
        (dataSource.dataSourceType == null ||
            dataSource.dataSourceType != dataSourceType)) return false;
    if (dataSourceId != null &&
        (dataSource.dataSourceId == null ||
            dataSource.dataSourceId != dataSourceId)) return false;
    if (platformType != null &&
        (dataSource.platformType == null ||
            dataSource.platformType != platformType)) return false;
    if (platformId != null &&
        (dataSource.platformId == null || dataSource.platformId != platformId))
      return false;
    if (platformAppType != null &&
        (dataSource.platformAppType == null ||
            dataSource.platformAppType != platformAppType)) return false;
    if (platformAppId != null &&
        (dataSource.platformAppId == null ||
            dataSource.platformAppId != platformAppId)) return false;
    if (applicationType != null &&
        (dataSource.applicationType == null ||
            dataSource.applicationType != applicationType)) return false;
    if (applicationId != null &&
        (dataSource.applicationId == null ||
            dataSource.applicationId != applicationId)) return false;
    return true;
  }

  factory MCDataSource.fromJson(Map<String, dynamic> json) {
    return MCDataSource._internal(
        dataSourceType: json['dataSourceType'],
        dataSourceId: json['dataSourceId'],
        platformType: json['platformType'],
        platformId: json['platformId'],
        platformAppType: json['platformAppType'],
        platformAppId: json['platformAppId'],
        applicationType: json['applicationType'],
        applicationId: json['applicationId'],
        dataFormat: MCDataFormat.fromString(json['dataFormat']?? "DATA"),
        dataSourceMetaData: json['dataSourceMetaData'] == null
            ? null
            : MCDataSourceMetaData.fromJson(json['dataSourceMetaData']),
        platformMetaData: json['platformMetaData'] == null
            ? null
            : MCPlatformMetaData.fromJson(json['platformMetaData']),
        platformAppMetaData: json['platformAppMetaData'] == null
            ? null
            : MCPlatformAppMetaData.fromJson(json['platformAppMetaData']),
        applicationMetaData: json['applicationMetaData'] == null
            ? null
            : MCApplicationMetaData.fromJson(json['applicationMetaData']),
        dataDescriptors: (json['dataDescriptors'] as List)
            .map((i) => MCDataDescriptor.fromJson(i))
            .toList());
  }

  Map<String, dynamic> toJson() => {
        'dataSourceType': dataSourceType,
        'dataSourceId': dataSourceId,
        'platformType': platformType,
        'platformId': platformId,
        'platformAppType': platformAppType,
        'platformAppId': platformAppId,
        'applicationType': applicationType,
        'applicationId': applicationId,
    'dataFormat': dataFormat==null?"DATA":dataFormat.toString(),
        'dataSourceMetaData': dataSourceMetaData,
        'platformMetaData': platformMetaData,
        'platformAppMetaData': platformAppMetaData,
        'applicationMetaData': applicationMetaData,
        'dataDescriptors': dataDescriptors,
      };


  static MCDataSource fromUUID(String uuid){
    String dataSourceType, dataSourceId, platformType, platformId, platformAppType, platformAppId, applicationType, applicationId;
    List<String> splits = uuid.split(_SEPARATOR);
    if (splits.length > 0 && splits[0] != null && splits[0].length != 0)
      dataSourceType = splits[0];
    if (splits.length > 1 && splits[1] != null && splits[1].length != 0)
      dataSourceId = splits[1];
    if (splits.length > 2 && splits[2] != null && splits[2].length != 0)
      platformType = splits[2];
    if (splits.length > 3 && splits[3] != null && splits[3].length != 0)
      platformId = splits[3];
    if (splits.length > 4 && splits[4] != null && splits[4].length != 0)
      platformAppType = splits[4];
    if (splits.length > 5 && splits[5] != null && splits[5].length != 0)
      platformAppId = splits[5];
    if (splits.length > 6 && splits[6] != null && splits[6].length != 0)
      applicationType = splits[6];
    if (splits.length > 7 && splits[7] != null && splits[7].length != 0)
      applicationId = splits[7];
    return MCDataSource._internal(dataSourceType: dataSourceType, dataSourceId: dataSourceId, platformType: platformType, platformId: platformId, platformAppType: platformAppType, platformAppId: platformAppId, applicationType: applicationType, applicationId: applicationId);
  }


  String getUUID() {
    return _prepString(dataSourceType) +
        _SEPARATOR +
        _prepString(dataSourceId) +
        _SEPARATOR +
        _prepString(platformType) +
        _SEPARATOR +
        _prepString(platformId) +
        _SEPARATOR +
        _prepString(platformAppType) +
        _SEPARATOR +
        _prepString(platformAppId) +
        _SEPARATOR +
        _prepString(applicationType) +
        _SEPARATOR +
        _prepString(applicationId);
  }

  String _prepString(String s) {
    if (s == null)
      return "";
    else
      return s;
  }

  @override
  List<Object> get props => [
    dataSourceType,
    dataSourceId,
    platformType,
    platformId,
    platformAppType,
    platformAppId,
    applicationType,
    applicationId,
    dataFormat.toString(),
    dataSourceMetaData,
    platformMetaData,
    platformAppMetaData,
    applicationMetaData,
    dataDescriptors
  ];
}
